enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,756 @@
+
+##### ControlPlaneMachineSet Synchronisation
+
+ControlPlaneMachineSets are not present in Cluster API.
+However, the ControlPlaneMachineSet was designed in such a way that we can extend the `template`` section (which is a discriminated union) to support Cluster API.
+
+A separate enhancement will be created for the support of Cluster API in the ControlPlaneMachineSet, however, for the sake of conversion,
+we expect the conversion to happen much in the same way as a regular MachineSet.
+
+ControlPlaneMachineSets (additionally to MachineSets) have a concept of failure domains.
+This is also a concept in Cluster API.
+When converting ControlPlaneMachineSets to Cluster API, we will ensure that the appropriate failure domain configuration is configured inside the Cluster API InfrastructureCluster,
+and that the failure domains from the ControlPlaneMachineSet Machine API configuration are reflected across.
+
+When deploying Machines using the future Cluster API extension, the InfrastructureTemplate will be combined with the named failure domains on the InfrastructureCluster
+to spread the Machines across multiple failure domains.
+
+#### Creation of new Machines
+
+When a Machine API MachineSet creates a new Machine, the synchronisation controller will create the equivalent Cluster API Machine based on the method described in [Machine Synchronisation](#machine-synchronisation).
+
+When a Cluster API MachineSet creates a new Machine, the synchronisation controller will create it’s Machine API equivalent if, and only if, the Cluster API MachineSet has a Machine API equivalent.
+This will allow users to switch a MachineSet back to Machine API at a later date if they wish to do so.
+
+When the Cluster API MachineSet has been created independently of Machine API, new Machines will not be created in Machine API to reflect the Machines owned by the Cluster API MachineSet.
+
+#### Deletion of Machines
+
+To enable the synchronisation of deletion, the synchronisation controller will use its own Finalizer on resources that have Machine API and Cluster API equivalents.
+
+This means that for a Machine, we expect the authoritative controller to perform the actual deletion work within the infrastructure provider.
+Once it has completed this effort, it will remove its own Finalizer.
+
+As soon as the synchronisation controller notices the Machine has been deleted, it should ensure that the Machine API/Cluster API equivalent (if present) is also deleted.
+It should then wait until the authoritative Machine controller removes its finalizer before removing both its own, and the Machine API/Cluster API Machine finalizer if present.
+In scenarios where both Machine API and Cluster API have been authoritative, it is expected that both Machine controllers will have added their own Finalizers, therefore we expect the synchronisation controller should be able to handle this.
```

Comment:
Fair point, I would lean to never having it based on some of the other context in this doc, though we would have to, if we are going to force remove it, ensure that the authoritative finalizer is in place before we remove it.

One thing I worry about here though, image I created the Machine as MAPI, and then switched over to CAPI. The syncer drops the finalizer from the MAPI object. Now, we delete the CAPI object, which in turns sets the deletiontimestamp on both sides.

Now, the CAPI resource controller is having difficulty removing the object, so we want to switch the authority back. The MAPI controller will not add its finalizer after the deletionTimestamp is set, that's common for controllers, so it will do nothing, as it doesn't think there's anything to do (not finalizer). And now we are blocked.
To mitigate this, the sync controller would have to add the finalizer, but, how do we know that this is the correct thing to do? It's possible that the finalizer was never present in the first place. So, now, we would have to always manage finalizer migration as part of the sync and authority handover?

I suspect it is safer to only interfere with the finalizers right at the end, when the authoritative API has concluded that it has removed the Machine.

Reply:
Again, IMHO the lesser evil is for the authoritative controller to ensure its finalizer is present until it has cleaned up the infrastructure, and then remove it.

```
if me.authoritative:
  ensureFinalizerPresent()
  // and carry on
else:
  ensureFinalizerAbsent()
  return nil

if deletionTimestamp != nil:
  ensureInfraScrubbed() || requeue
  ensureFinalizerRemoved()
```

Reply:
I don't think it's that simple. If one controller recognises that it isn't authoritative and removes its finalizer before the authoritative one adds its own finalizer, then theres a period where neither has a finalizer and that means we risk leaking resources if a well timed delete happens

If we are concerned about the UX, I think it would be safer to just ensure that if the authoritative resource has its finalizer, then the sync controller would add the finalizer to the non-authoritative as well

Reply:
We have made some movement on the hand over mechanism, I believe the handover mechanism can handle the concern over the finalizers here, will review and update this section